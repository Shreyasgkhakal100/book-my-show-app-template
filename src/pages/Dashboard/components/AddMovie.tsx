import axios from 'axios';
import React, { FC, useState } from 'react';
import { AddMovieProps, Movie } from './types';

export const AddMovie: FC<AddMovieProps> = ({ getMovies }) => {
    const [movie, setMovie] = useState<Movie>({ title: '', duration: 0 });

    function handleChange(e: React.FormEvent<HTMLInputElement>): void {
        const { value, id } = e.currentTarget;
        setMovie({ ...movie, [id]: value });
    }

    async function handleSubmit() {
        try {
            await axios.post('/api/movies', movie);
            getMovies();
        } catch (err) {
            console.log(err);
        }
    }
    return (
        <form className="form-inline">
            <div className="form-group mx-sm-3 mb-2">
                <label htmlFor="title" className="sr-only">
                    Title
                </label>
                <input
                    type="text"
                    className="form-control"
                    id="title"
                    placeholder="Title"
                    onChange={e => {
                        handleChange(e);
                    }}
                    minLength={1}
                />
            </div>
            <div className="form-group mx-sm-3 mb-2">
                <label htmlFor="duration" className="sr-only">
                    Duration
                </label>
                <input
                    type="number"
                    className="form-control"
                    id="duration"
                    min="300"
                    max="21600"
                    placeholder="Duration"
                    onChange={e => {
                        handleChange(e);
                    }}
                />
            </div>
            <button
                type="submit"
                className="btn btn-primary mb-2"
                onClick={e => {
                    e.preventDefault();
                    handleSubmit();
                }}
            >
                Add Movie
            </button>
        </form>
    );
};
