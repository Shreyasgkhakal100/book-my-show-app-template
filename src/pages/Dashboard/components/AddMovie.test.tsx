import { mockAxios, renderWithRouter } from '@utils/test-utils';
import { AddMovie } from './AddMovie';
import { fireEvent, screen, waitFor } from '@testing-library/react';

describe('AddMovies', () => {
    let mockGetMovies;

    beforeEach(() => {
        mockGetMovies = jest.fn();
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('should render properly', () => {
        const { container } = renderWithRouter(<AddMovie getMovies={mockGetMovies} />);
        expect(container).toMatchSnapshot();
    });

    it('should call mockGetMovies at onClick event at button if post request is complete', async () => {
        mockAxios.onPost('/api/movies').reply(200);
        renderWithRouter(<AddMovie getMovies={mockGetMovies} />);
        const button = screen.getByText('Add Movie');
        fireEvent.click(button);
        await waitFor(() => {
            expect(mockGetMovies).toHaveBeenCalled();
        });
    });

    it('should not call mockGetMovies at onClick event at button if post request sends error', async () => {
        mockAxios.onPost('/api/movies').reply(400);
        renderWithRouter(<AddMovie getMovies={mockGetMovies} />);
        const button = screen.getByText('Add Movie');
        fireEvent.click(button);
        await waitFor(() => {
            expect(mockGetMovies).not.toHaveBeenCalled();
        });
    });
});
