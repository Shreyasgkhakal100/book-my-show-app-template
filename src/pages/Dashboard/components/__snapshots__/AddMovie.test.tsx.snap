// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddMovies should render properly 1`] = `
<div>
  <form
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3 mb-2"
    >
      <label
        class="sr-only"
        for="title"
      >
        Title
      </label>
      <input
        class="form-control"
        id="title"
        placeholder="Title"
        type="text"
      />
    </div>
    <div
      class="form-group mx-sm-3 mb-2"
    >
      <label
        class="sr-only"
        for="duration"
      >
        Duration
      </label>
      <input
        class="form-control"
        id="duration"
        min="5"
        placeholder="Duration"
        type="number"
      />
    </div>
    <button
      class="btn btn-primary mb-2"
      type="submit"
    >
      Add Movie
    </button>
  </form>
</div>
`;
